### DISCUSSION TITLE
Help Installing OpenFast

### DISCUSSION THREAD

**Discussion by unistudent2018** [OpenFAST/openfast#1316] [General]
Hello, I have followed https://www.youtube.com/watch?v=h_-Qe2m7rN0 on how to install OpenFast, but when i come to the Build Solution I get a tonne of errors 
![image](https://user-images.githubusercontent.com/117730294/200547800-70df8c91-6143-43ab-824b-7b6cbc6d902c.png)
![image](https://user-images.githubusercontent.com/117730294/200547849-f5ff2105-56af-490c-a43d-bc948aee6918.png)
Does anyone have any ideas where I went wrong or can help me? I am installing it as a windows instsallation.

**Comment by jjonkman** [comment]
Dear @unistudent2018,

Can you share what is reported in the "Output" window (just to the right of the "Error List" window)?

Best regards,

**Reply by unistudent2018** [reply]
This is what is showing in the Output window

![image](https://user-images.githubusercontent.com/117730294/200653291-49508de0-927c-4e31-af28-226fc596e7ed.png)

**Reply by ebranlard** [reply]
- Do you know which version of OpenFAST you downloaded? If you did a "git clone", when you can type `git describe` in the openfast directory, this will give us the exact version that you are using (just in case there is an issue). 

- It's possible that you need to update the "Platform toolset" in Visual Studio. Do to that, in the Solution Explorer:  
    - Right Clik on "MAP_dll", >  Properties  > Configuration Properties > General > Platform Toolset: Select the top option (on my machine it's "Visual Studio 2019 v142).
    - Do the same steps for "FAST_Registry"

- You can try building "MAP_dll" and "FAST_Registry" individually, making sure those work. (Right click on them > Build)  

Let us know if that helps

**Reply by jjonkman** [reply]
Also, it is typically the first error shown in the "Output" window that is most informative.

Best regards,

**Reply by unistudent2018** [reply]
Hello,

I have version v2.1.0-3324-g5f3fb6ef from doing the git describe
I have updated the MAP.dll and FAST_Registry, they were both set to 2015 (not installed) they are now 2022 v143

That has given me these errors:
![image](https://user-images.githubusercontent.com/117730294/200795219-f539d3bf-a424-47de-b3ce-90f9964ada18.png)

And in the output window it says:
========== Build: 1 succeeded, 3 failed, 0 up-to-date, 1 skipped ==========


Bit of a daft question maybe. Can i move the whole openfast directory to a different location? It is currently in my c drive but i'm running out of space there where I have a tonne of space on my d drive. Can I just copy and paste it?

**Reply by jjonkman** [reply]
Again, please share the contents of the "Output" window, especially around the first error reported.

**Reply by ebranlard** [reply]
Yes, you can move the entire OpenFAST folder, and redo the compilation. It could be that you lacked disk space, or you changed the solution configuration somehow. 
I would advise the following: 
- Make sure you don't change the "Solution Configuration" in between compilations (it seems you are using Debug x64). 
- Build > Clean solution
- Right click on FAST_Registry > Build . Make sure this is successful. If not, most likely you need to redo the steps I mentioned above, and update the platform toolset so that it doesn't say "Visual Studio 2015" anymore.
- Right click on MAP_dll > Build . Make sure this is successful.
- Right click on FASTlib > Build. Make sure this is successful
- Right click on FAST > Build. Make sure this is successful
If that works, in the future, you can probably just do  "Build> Build Solution"

**Reply by unistudent2018** [reply]
Ok I did build clean solution. That was successful.
I then did the FAST_Registry Build. That was successful
then MAP_Dll, that failed. This is the output window
![image](https://user-images.githubusercontent.com/117730294/200957382-dc4007e1-a6a0-4ebf-8553-df27f05e2d80.png)
then I tried FASTlib Build. That came up with 4828 errors and 751 warnings. the Output window and error list are really too big to screenshot.
then I tried FAST. That had 3 errors and 13 warnings.
![image](https://user-images.githubusercontent.com/117730294/200958731-9c813d30-ec7b-4aba-b3df-e08496cfa656.png)
![image](https://user-images.githubusercontent.com/117730294/200958767-a07f27d5-bc8b-4cca-a135-fa220499c545.png)

**Reply by jjonkman** [reply]
Regarding the SDK error, a similar question was asked and answered in the following issue: https://github.com/OpenFAST/openfast/issues/787.  Does that solve your problem?

**Comment by unistudent2018** [comment]
Hi, yes that fixed the SDK error. 
I now get just 1 error and 171 warnings
![image](https://user-images.githubusercontent.com/117730294/201037247-cae13f33-5881-4ca8-b958-28b8cb699d01.png)
![image](https://user-images.githubusercontent.com/117730294/201037273-71743ed4-98b1-4e6e-89b0-e3b30527c19f.png)

**Comment by unistudent2018** [comment]
I have moved it to my D drive to get some more space.
I went through these steps again

- Build > Clean solution - **successful**
- FAST_Registry > Build - **successful**
- MAP_dll > Build - **successful**
- FASTlib > Build - **successful**
- FAST > Build - **unsuccessful**
![image](https://user-images.githubusercontent.com/117730294/201039008-b30b9f0e-b109-4d35-87fd-d4c4ccdc6cdc.png)
![image](https://user-images.githubusercontent.com/117730294/201039046-395ed96f-3dab-4015-8dff-9d236bb39aa2.png)

**Reply by jjonkman** [reply]
Regarding the MKL error, a similar question was asked and answered in the following issue: https://github.com/OpenFAST/openfast/issues/1249.  Does that solve your problem?

**Comment by unistudent2018** [comment]
I looked at that and think I fixed it from the link, thanks.
I carried out all the builds and got this
![image](https://user-images.githubusercontent.com/117730294/201198556-f6618d72-83c2-4bf9-bc3a-3e9785e52ff4.png)
Bit confused because it said it failed but there is no warnings and no errors?
Is that it working correctly?

**Reply by jjonkman** [reply]
My guess is that you didn't fix the MKL error.  I would suggest closing Visual Studio, opening it up again and start build from scratch and see if you see an error.

**Comment by unistudent2018** [comment]
Tried it with a new visual studio and got this message. I just tried it as Build Solution
![image](https://user-images.githubusercontent.com/117730294/201444161-39a3c4dd-6429-49e6-aa50-cc55f329067f.png)

**Reply by jjonkman** [reply]
It looks like there is an error in the linking stage.  Do you have an error showing up in the Error List?  I'm guessing you are still missing a library.

**Comment by unistudent2018** [comment]
there are no errors showing.

**Reply by jjonkman** [reply]
Did the file you want (OpenFAST executable) get created?

If not, what is written to the _BuildLog.htm_?

**Comment by unistudent2018** [comment]
Tried it again and this was the screen
![image](https://user-images.githubusercontent.com/117730294/201490121-2c3f0de1-e450-4818-ac04-79a91cdc5c9c.png)

This is what is written in the BuildLog

Linking...
Link /OUT:"..\..\build\bin\\openfast_x64_Debug.exe" /INCREMENTAL:NO /NOLOGO /MANIFEST:NO /DEBUG /PDB:"D:\Program Files\openfast\build\bin\openfast_x64_Debug.pdb" /SUBSYSTEM:CONSOLE /STACK:9999999 /IMPLIB:"D:\Program Files\openfast\build\bin\openfast_x64_Debug.lib" -qm64 /qoffload-ldopts="-mkl=sequential" "x64\Debug\FAST_Prog.obj" "D:\Program Files\openfast\build\bin\FASTlib_x64_Debug.lib"
FASTlib_x64_Debug.lib(mapapi.obj) : warning LNK4075: ignoring '/EDITANDCONTINUE' due to '/INCREMENTAL:NO' specification
LINK : fatal error LNK1104: cannot open file 'mkl_intel_lp64.lib'


FAST - 0 error(s), 0 warning(s)


Where would I see the OpenFast executable? In the parent folder? If so then no it isnt there.
![image](https://user-images.githubusercontent.com/117730294/201490213-201e742c-1931-4edb-a15c-dd953bca778f.png)

**Reply by jjonkman** [reply]
It sounds like you never solved the MKL issue.  You'll need to install the MKL library per #1249.  Once compiled, the OpenFAST executable should exist in the _build/bin_ directory.

**Comment by unistudent2018** [comment]
![image](https://user-images.githubusercontent.com/117730294/201526962-74545c14-a9fa-4852-9343-91d141eeb736.png)
Is that it? Is it now working??? Say it aint so! lol
![image](https://user-images.githubusercontent.com/117730294/201526999-7b46d73a-960d-44dd-a5a2-a9151b4f0665.png)
Is that where the OpenFast executable should be?

**Reply by jjonkman** [reply]
It looks like you've only compiled the FAST registry and MAP++ DLL, not the OpenFAST executable.  The other projects are marked "incompatible".  Regarding the "incompatible" projects, see a similar question asked and answered in the following issue: https://github.com/OpenFAST/openfast/issues/328#issuecomment-642902611.

**Reply by unistudent2018** [reply]
Hi, I see the question asked and I see the answer but the answer is a little incomplete for me as they say when they right click on it it should be obvious what one to select, but i dont see anything obvious that I have not tried.

**Comment by rafmudaf** [comment]
@unistudent2018 Is it required that you compile the code yourself? This is typically needed when you have or will change the source code or you're using uncommon hardware. If you will not be modifying the source code and you are using a common computer, you can avoid the challenges in compiling by downloading a binary file from GitHub Release (see the [latest release](https://github.com/OpenFAST/openfast/releases/tag/v3.3.0)).

**Reply by unistudent2018** [reply]
It might be better if I just tell you what I am trying to do.
We are trying to re-design an Offshore 15MW wind turbine platform. So we want to use the reference platform that is in OpenFast, test it for their environmental conditions, then essentially "move" that turbine to our environmental conditions and see where it fails/breaks/needs improvement.
Then redesign the platform, and test again showing how wonderfully awesome we are that we have made a floating offshire wind turbine platform that can handle 10m waves and high winds.

Currently we have me trying to get OpenFast to work. my colleague who has tried and given up on OpenFast but has got Fast v8 working, but the 15MW reference platform doesnt easily port over to Fast v8 so we're kind of a little stuck right now.

**Reply by jjonkman** [reply]
Dear @unistudent2018,

I don't see anything in your message that states why you need to compile OpenFAST.  If you are using Windows, you can use the precompiled Windows binary executable of OpenFAST as @rafmudaf indicated.  You only need to compile OpenFAST if you are not using Windows or if you need to change the source code.

Best regards,

**Reply by rafmudaf** [reply]
> You only need to compile OpenFAST if you are not using Windows

Precompiled binaries are actually available for all platforms (Windows, macOS, Linux), so there is no requirement to compile the code in order to use it for any system.

For macOS and Linux, OpenFAST can be installed through conda via conda-forge. This also works on Windows using the Windows Subsystem for Linux (WSL). For native Windows, the binaries are distributed through GitHub Releases.

Detailed descriptions of these procedures are available in the [docs](https://openfast.readthedocs.io/en/main/source/install/index.html#download-binaries).

**Reply by jjonkman** [reply]
Thanks for correcting me @rafmudaf.  As a Windows user, I forget the conda option :-).

**Reply by unistudent2018** [reply]
So that is what I initially tried to follow the first time, but i'll be honest, it is (still) quite confusing (to me anyway).
the OpenFAST v3.3.0 doesnt have a zip file, so I am unsure what to actually download? All files under Assets? Do I need the source code zip?
And is that all I need to download? I don't follow any of the other steps?

**Reply by rafmudaf** [reply]
Ah @unistudent2018 its true that the documentation is inaccurate. Previously, we bundled all the binaries into a zip file, but lately we've been posting them individually to the GitHub page. So now the contents of the zip file described in the documentation exists directly in the Assets list. Does this make it more clear?

**Reply by unistudent2018** [reply]
A little. So do I just download everything in the assets list? Do i need to keep them in a specific directory? Do I need to do anything else after I have done that?

**Reply by jjonkman** [reply]
From your work outline, it sounds like you need the _OpenFAST_x64.exe_ (to simulate with OpenFAST), and probably the _TurbSim_x64.exe_ (for generating inflow).  You can place these files where you want.  To run OpenFAST on Windows, the syntax is described in the following issue: https://github.com/OpenFAST/openfast/issues/434#issuecomment-670513966/.

**Reply by unistudent2018** [reply]
If I dont need all the other stuff, do I need to have the Intel oneAPI Base Toolkit, Fortran compiler, etc?

**Comment by unistudent2018** [comment]
Based on the comments above saying I don't need to compile OpenFast I thought I would give this video another go.
https://www.youtube.com/watch?v=I8TP84swWZE&t=123s
And I felt like I could now follow the steps and understand what is happening/what I am doing. However, when I tried to run it like in the steps he shows at the 7min mark (for his it all works) I get this:
![image](https://user-images.githubusercontent.com/117730294/202417164-dcd2f947-7ba8-4fd7-91d8-156cc4b6e290.png)
I honestly thought I had figured it out but FATAL ERROR does not seem good.

**Reply by unistudent2018** [reply]
Think I got it. I was using the wrong exe file as I was trying to copy him exactly.
![image](https://user-images.githubusercontent.com/117730294/202420746-c2018e26-1964-4153-bf57-d4bdf11857f7.png)

**Reply by bjonkman** [reply]
Your error when using the executable for v3.0.0 is that you were using input files for version 3.3.0. As long as the input files match the version of the executable, you should be fine.
